client/webpack-utils/extract-manifest-plugin.js

```diff
@@ -0,0 +1,106 @@
+		this.options = {
+			manifestName: 'manifest',
+			runtimeChunk: 'runtime',
+			globalManifest: 'window.__WEBPACK_MANIFEST',
+			...options,
+		};
+	}
+
+	apply( compiler ) {
+		// Tapping into compiler.make allow us to come after the regular TemplatedPathPlugin
+		compiler.hooks.compilation.tap( PLUGIN_NAME, ( compilation ) => {
+			const globalManifest = this.options.globalManifest;
+			const manifestContent = [ `${ globalManifest }=new Map();` ];
+			const quoteRegex = new RegExp( /^(['"])/ );
+			const variableRegex = new RegExp( /\[(.*?)\]/gi );
+			const variablesProcessed = new Set();
+
+			// This method is hock several times. First, it is called to generate the filename of each chunk. In that
+			// case, `data.hashWithLength` won't be defined and we just forward the output of the original filenameExpression.
+			//
+			// It will also be called to generate an expression that receives a chunkId and returns the filename where that chunkId is
+			// saved, something like `({1:'chunk1', 2:'chunk2'}[chunkId])+".min.js"` (simplified example). In this case, capture the expressions
+			// that do the variable replacement and store them in a global Map. This allows reusing those expressions later in the runtime so we
+			// don't have to concatenate the chunkId map multiple times.
+			//
+			// Simplified example: if filenameExpression is "[name].[chunkhash].js", we end up with
+			//
+			// - In runtime:
+			//		function jsonpScriptSrc(chunkId) {
+			//			return window.__WEBPACK_MANIFEST.get('name')(chunkId) + "." + window.__WEBPACK_MANIFEST.get('chunkhash')(chunkId) + ".js"
+			//		}
+			//
+			// - In manifest:
+			//		window.__WEBPACK_MANIFEST=new Map();
+			//		window.__WEBPACK_MANIFEST.set('name', (chunkId) => {1:'my-name'}[chunkId]);
+			//		window.__WEBPACK_MANIFEST.set('chunkhash', (chunkId) => {1:'abcd1234'}[chunkId]);
+			//
+			// So when webpack calls `jsonpScriptSrc(1)` it gets `my-name.abcd1234.js`
+			compilation.mainTemplate.hooks.assetPath.tap( PLUGIN_NAME, ( filenameExpression, data ) => {
+				// When there is no `hashWithLength`, this hook has been called to generate the file name of a chunk.
+				// Return whatever Webpack generates unmodified.
+				if ( ! data.hashWithLength || ! ( data.chunk && data.chunk.hashWithLength ) ) {
+					return filenameExpression;
+				}
+
+				// Capture the quote to make sure we are using the right one when concatenating code.
+				const [ , quote = '"' ] = filenameExpression.match( quoteRegex );
+
+				return filenameExpression.replace( variableRegex, function ( match, variable ) {
+					// If the variable has been prcoessed (i.e. is already in the manifest), dont' try to get the
+					// value again.
+					if ( ! variablesProcessed.has( variable ) ) {
+						// Call webpack again to generate the chunk map for that specific variable.
+						const nameExpression = compilation.mainTemplate.getAssetPath( `"[${ variable }]"`, {
+							chunk: data.chunk,
+						} );
```

Comment:
Forwarding _all_ variables, instead of just a curated subset, can be fragile. Also, not sure if forwarding just the `data.chunk` option, instead of the entire `data` object, is 100% correct.

I can break this plugin by using the `[contenthash]` variable in the chunk name template. For `contenthash` to work, we need to forward not only the `chunk` option, but also `contentHashType` (which usually has value `"javascript"`).

Reply:
Good point, I changed it to pass the full `data` and `assetInfo` objects to mimic the original function. Also added a try/catch to skip the replacement if there is any problem getting the chunk map replacer function.

Tested it with all allowed variables in Webpack 4 (based on https://github.com/webpack/webpack/blob/webpack-4/lib/TemplatedPathPlugin.js), with a name like `[name]-[chunkhash:11]-[contenthash:3]-[id]-[name]-[query].min.js` and it seems to work. It also works with `[hash]` (but webpack doesn't allow me to use `[hash]` and `[chunkhash]` at the same time, not related to this plugin). Variables `moduleid`, `file`, `modulehash`, `filebase` and `url` also are not allowed by webpack  as `output.filename` or `output.chunkFilename` in our config (looking at the source looks like those unallowed variables are used to generate sourcemap)

I think it is safer to allow passing any variable between `[`,`]` and not only a fixed set. This will make the plugin work even if an external plugin adds a new variable, for example. In theory if `getAssetPath( "...[var]...", options, assetInfo )` works, `getAssetPath("[var]",options, assetInfo)` should always work, independently of what `var` is.

Reply:
> Variables `moduleid`, `file`, `modulehash`, `filebase` and `url` also are not allowed by webpack

They are probably related to some exotic library targets, not used for traditional app bundles. Interesting that I couldn't find any usage of the `module*` stuff in the webpack codebase, i.e., any `assetPath` call that would pass `module` as an option.